File: jeph2sworm-core/jeph2sworm/llm/mistral_provider.py
"""Mistral Provider - Mistral Large, Codestral, Mistral Small."""

from __future__ import annotations

from typing import List

import structlog

from jeph2sworm.llm.base_provider import BaseProvider, LLMMessage, LLMResponse

logger = structlog.get_logger()

MODELS = [
    "mistral-large-latest",
    "codestral-latest",
    "mistral-small-latest",
    "mistral-medium-latest",
]

TASK_BEST = {
    "coding": "codestral-latest",
    "planning": "mistral-large-latest",
    "reasoning": "mistral-large-latest",
    "design": "mistral-large-latest",
    "testing": "mistral-small-latest",
    "general": "mistral-large-latest",
}


class MistralProvider(BaseProvider):
    provider_name = "mistral"
    supports_function_calling = True
    supports_vision = False
    max_context_window = 128_000

    def available_models(self) -> List[str]:
        return MODELS

    def best_model_for(self, task: str) -> str:
        return TASK_BEST.get(task, "mistral-large-latest")

    async def complete(
        self,
        messages: List[LLMMessage],
        model: str,
        temperature: float = 0.7,
        max_tokens: int = 4096,
        **kwargs,
    ) -> LLMResponse:
        import litellm

        response = await litellm.acompletion(
            model=f"mistral/{model}" if not model.startswith("mistral/") else model,
            messages=[m.model_dump() for m in messages],
            temperature=temperature,
            max_tokens=max_tokens,
            api_key=self.api_key,
            **kwargs,
        )

        choice = response.choices[0]
        usage = dict(response.usage) if response.usage else {}

        return LLMResponse(
            content=choice.message.content or "",
            model=model,
            provider=self.provider_name,
            usage=usage,
            finish_reason=choice.finish_reason or "stop",
        )

File: jeph2sworm-vscode/src/providers/mistral-provider.ts
/**
 * Mistral Provider — proxy to Mistral Large / Medium / Small models.
 */

import { BaseProvider, LLMMessage, LLMResponse, ProviderConfig } from './base-provider';

export class MistralProvider extends BaseProvider {
  readonly name = 'mistral';
  readonly models = ['mistral-large-latest', 'mistral-medium-latest', 'mistral-small-latest', 'codestral-latest'];

  constructor(config: ProviderConfig) {
    super({ baseUrl: 'https://api.mistral.ai/v1', ...config });
  }

  async complete(messages: LLMMessage[], model?: string): Promise<LLMResponse> {
    const chosen = model ?? this.config.defaultModel ?? 'mistral-large-latest';
    const res = await fetch(`${this.config.baseUrl}/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this.config.apiKey}`,
      },
      body: JSON.stringify({ model: chosen, messages }),
    });
    const data = await res.json() as any;
    return {
      content: data.choices?.[0]?.message?.content ?? '',
      model: chosen,
      tokensUsed: data.usage?.total_tokens ?? 0,
      finishReason: data.choices?.[0]?.finish_reason ?? 'unknown',
    };
  }

  bestModelFor(task: 'coding' | 'planning' | 'design' | 'quick'): string {
    switch (task) {
      case 'coding':
        return 'codestral-latest';
      case 'planning':
        return 'mistral-large-latest';
      case 'design':
        return 'mistral-medium-latest';
      case 'quick':
        return 'mistral-small-latest';
    }
  }
}
